8334900: IOOBE when adding data to a Series of a BarChart that already contains data

drmarmac opened the pull request:
This PR is a fix for another IOOBE that I discovered while working on #1476.

The PR simplifies the code for adding a series that already contains data by adding the data points one-by-one.
As far as I can see no attempt was previously made to optimize the bulk operation except for some trivial O(1) operations, so this should have no noticable performance impact.

Accidentally this fixes another bug related to the missing "negative" style class when negative data values are added.

Also, the PR aligns the handling of duplicate categories with the behavior clarified in #1476, when there are duplicates in the data that was already in the series before the series was added to the chart.

Note a change was made to the createTestSeries() test method, letting it start at index 1, avoiding the duplicate data items resulting from multiplying by 0.
Without this change `testSeriesRemoveAnimatedStyleClasses` would fail because it counts the number of plot children, where the duplicates are now removed.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8334900](https://bugs.openjdk.org/browse/JDK-8334900): IOOBE when adding data to a Series of a BarChart that already contains data (**Bug** - P4)
 * [JDK-8334901](https://bugs.openjdk.org/browse/JDK-8334901): Style class "negative" is not always added for new BarChart data with negative value (**Bug** - P4)


### Reviewers
 * [Michael Strauß](https://openjdk.org/census#mstrauss) (@mstr2 - Committer)
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1488/head:pull/1488` \
`$ git checkout pull/1488`

Update a local copy of the PR: \
`$ git checkout pull/1488` \
`$ git pull https://git.openjdk.org/jfx.git pull/1488/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1488`

View PR using the GUI difftool: \
`$ git pr show -t 1488`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1488.diff">https://git.openjdk.org/jfx/pull/1488.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1488#issuecomment-2187504044)

drmarmac commented:
/issue add JDK-8334901

mstr2 reviewed (approved):
Good idea to deduplicate the code in `dataItemAdded` and `seriesAdded`. The patch looks good.

kevinrushforth commented:
@drmarmac You need to merge the latest upstream master and fix the merge conflicts before this can proceed.

Reviewers: @andy-goryachev-oracle @mstr2 

/reviewers 2

drmarmac commented:
I've updated to the current master branch.

andy-goryachev-oracle commented:
I don't see the 'negative' style added to negative bars.  I do see BarChart:261 being hit, but nothing in the scene graph.

Tested with the Monkey Tester and double checked with Scenic View 11.0.2:

![Screenshot 2024-07-12 at 09 18 36](https://github.com/user-attachments/assets/9f7d3a23-cce2-4d36-91e0-e84f24e0fd49)

drmarmac commented:
good catch, that was another situation where the "negative" style class was not added when adding a series that already contains data (which is the situation when the IOOBE occured).
I fixed this in the relevant place, did some more code deduplication, and added another unit test.

andy-goryachev-oracle reviewed (commented):
The changes look good.

I think it would be nice to test all possible ways the `negative` class is added or removed as part of this PR.  What do you think?

drmarmac commented:
I've added asserts verifying that the "negative" style class is added to the relevant tests, plus a new test to assert it is removed when data is changed.

drmarmac commented on modules/javafx.controls/src/test/java/test/javafx/scene/chart/BarChartTest.java:
```diff
@@ -276,6 +275,69 @@ public void testAddingMultipleSeriesWithDuplicateCategories() {
+        startApp();
+        BarChart<String, Number> chart = new BarChart<>(new CategoryAxis(), new NumberAxis());
+        XYChart.Series<String, Number> series = new XYChart.Series<>();
+        chart.getData().add(series);
```
good idea, done

drmarmac commented:
> I am going to ask this question - unrelated to this PR, but in the context of this PR.
> 
> Consider the case of a context menu, where the application needs to know the data point from the screen/local coordinates.
> 
> Currently, we can get the Node from the Data instance (XYChart.Data.node property), but not the other way around - that is, given the chart, and given the PickResult of a MouseEvent, it is impossible to get the data point corresponding to a Node. In other words, we have the model->view path but not the view->model one.
> 
> In theory, the app developer could monitor the `node` property of Data and use `Node.getProperties()` map to set the Data instance on the node, or they could iterate through all the visible data points, matching the node with the data point, so it is not a big issue.
> 
> I am interested in your opinion. What do you think?

Good question, I'd say it is easy enough to loop through the chart's data items to find the correct node in that case, so I don't see a need to add this on the jfx side.
Also, currently Data.getNode() just returns an opaque Node, so I don't see a good place where to expose that without either changing that API. Adding an Node.getProperties() entry should probably be done only by application developers.

andy-goryachev-oracle reviewed (approved):
thank you for making the changes!  looks good now.

drmarmac commented:
/integrate
